File: src/NodeParser/UnionNodeParser.ts
import ts from "typescript";
import { Context, NodeParser } from "../NodeParser.js";
import { SubNodeParser } from "../SubNodeParser.js";
import { UnionType } from "../Type/UnionType.js";
import { BaseType } from "../Type/BaseType.js";
import { notNever } from "../Utils/notNever.js";
import { NeverType } from "../Type/NeverType.js";
import { notUndefined } from "../Utils/notUndefined.js";

export class UnionNodeParser implements SubNodeParser {
    public constructor(
        protected typeChecker: ts.TypeChecker,
        protected childNodeParser: NodeParser,
    ) {}

    public supportsNode(node: ts.UnionTypeNode): boolean {
        return node.kind === ts.SyntaxKind.UnionType;
    }

    public createType(node: ts.UnionTypeNode, context: Context): BaseType {
        const types = node.types
            .map((subnode) => {
                return this.childNodeParser.createType(subnode, context);
            })
            .filter(notNever)
            .filter(notUndefined);

        if (types.length === 1) {
            return types[0];
        } else if (types.length === 0) {
            return new NeverType();
        }

        return new UnionType(types);
    }
}

File: internal/parser/union_node_parser.go
package parser

import (
	"github.com/microsoft/typescript-go/shim/ast"
	"github.com/microsoft/typescript-go/shim/checker"

	"github.com/vega/ts-json-schema-generator/internal/types"
)

// UnionNodeParser parses union type nodes (src/NodeParser/UnionNodeParser.ts).
type UnionNodeParser struct {
	typeChecker     *checker.Checker
	childNodeParser NodeParser
}

func NewUnionNodeParser(typeChecker *checker.Checker, childNodeParser NodeParser) *UnionNodeParser {
	return &UnionNodeParser{typeChecker: typeChecker, childNodeParser: childNodeParser}
}

func (p *UnionNodeParser) SupportsNode(node *ast.Node) bool {
	return node.Kind == ast.KindUnionType
}

func (p *UnionNodeParser) CreateType(node *ast.Node, ctx *Context, _ *types.ReferenceType) types.Type {
	var memberTypes []types.Type
	for _, subnode := range node.AsUnionTypeNode().Types.Nodes {
		typ := p.childNodeParser.CreateType(subnode, ctx, nil)
		if types.NotNever(typ) {
			memberTypes = append(memberTypes, typ)
		}
	}

	if len(memberTypes) == 1 {
		return memberTypes[0]
	} else if len(memberTypes) == 0 {
		return &types.NeverType{}
	}

	return types.NewUnionType(memberTypes)
}
